### PR TITLE
feat(core/vm): revert modexp precompiled contract if input is not u256

### PR DESCRIPTION
**Pull request by NazariiDenha**
## 1. Purpose or design rationale of this PR
`modexp` precompile supports arbitrary-length input which makes it hard to support it in the zkEVM. But it seems that most contracts that use it provide a uint256 input. It seems sufficient for us to support this use case and revert otherwise.

Team discussion: https://scrollco.slack.com/archives/C03TK1LFBB8/p1685676526484859

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [x] Yes

### CODE REVIEW

File: core/vm/contracts.go [hunk 1]
```diff
@@ -410,6 +410,10 @@ func (c *bigModExp) Run(input []byte) ([]byte, error) {
 		expLen  = new(big.Int).SetBytes(getData(input, 32, 32)).Uint64()
 		modLen  = new(big.Int).SetBytes(getData(input, 64, 32)).Uint64()
 	)
+	// Check that all inputs are `u256` (32 - bytes), revert otherwise
+	if baseLen != 32 || expLen != 32 || modLen != 32 {
+		return nil, ErrExecutionReverted
```
Comment (NazariiDenha) [inline core/vm/contracts.go]: Should we return `ErrExecutionReverted` or some specific error as in https://github.com/scroll-tech/go-ethereum/pull/337

Comment (Thegaram) [reply core/vm/contracts.go]: As far as I know, the difference is that `ErrExecutionReverted` uses up all the gas provided to the call context, while other errors do not.

https://github.com/scroll-tech/go-ethereum/blob/8057f78edf121bba83b9f503bb6d1073d12f1d41/core/vm/evm.go#L244-L246

I think a modexp-specific error might be better here. What do you think? @zimpha

**Review by Thegaram — COMMENTED** [review]
LGTM, please bump version (patch is probably enough since this is a small change and we haven't released v4 yet so we can group multiple breaking changes togethrer).

File: core/vm/contracts.go (hunk 1 above)
Comment (NazariiDenha) [reply core/vm/contracts.go]: >  `ErrExecutionReverted` uses up all the gas provided to the call context, while other errors do not.

Yes, that is why I am asking

Comment (zimpha) [reply core/vm/contracts.go]: Since we don't provide a reason for the revert, it's better to return a specific error.

File: core/vm/contracts.go [hunk 2]
```diff
@@ -410,6 +411,10 @@ func (c *bigModExp) Run(input []byte) ([]byte, error) {
 		expLen  = new(big.Int).SetBytes(getData(input, 32, 32)).Uint64()
 		modLen  = new(big.Int).SetBytes(getData(input, 64, 32)).Uint64()
 	)
+	// Check that all inputs are `u256` (32 - bytes), revert otherwise
+	if baseLen != 32 || expLen != 32 || modLen != 32 {
```
Comment (Thegaram) [inline core/vm/contracts.go:415]:
```suggestion
	if baseLen > 32 || expLen > 32 || modLen > 32 {
```

**Comment by Thegaram** [comment]
Can you please also add some tests?

File: core/vm/contracts.go [hunk 3]
```diff
@@ -35,7 +35,8 @@ import (
 )
 
 var (
-	errPrecompileDisabled = errors.New("sha256, ripemd160, blake2f precompiles temporarily disabled")
+	errPrecompileDisabled     = errors.New("sha256, ripemd160, blake2f precompiles temporarily disabled")
+	errModexpUnsupportedInput = errors.New("modexp temporarily accepts only 32-byte (256-bit) input")
```
Comment (icemelon) [inline core/vm/contracts.go]:
```suggestion
	errModexpUnsupportedInput = errors.New("modexp temporarily accepts only 32-byte (256-bit) inputs")
```

**Comment by lispc** [comment]
we think supporting less than 32byte input is viable but requires non trivial work (while brings little benefits since all dapps i found use u256 and it is more convenient for solidity developer to directly use u256 than building u160/u64/... inputs themselves). So we prefer only supporting u256.

File: core/vm/contracts.go (hunk 2 above)
Comment (Thegaram) [reply core/vm/contracts.go:415]: Resolving in light of Zhuo's request.

File: core/vm/testdata/precompiles/modexp.json [hunk 4]
```diff
@@ -1,121 +1,16 @@
...
+    "Input": "0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2efffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f",
     "Expected": "0000000000000000000000000000000000000000000000000000000000000000",
     "Name": "eip_example2",
     "Gas": 13056,
     "NoBenchmark": false
-  },
-  {
-    "Input": "000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000040e09ad9675465c53a109fac66a445c91b292d2bb2c5268addb30cd82f80fcb0033ff97c80a5fc6f39193ae969c6ede6710a6b7ac27078a06d90ef1c72e5c85fb502fc9e1f6beb81516545975218075ec2af118cd8798df6e08a147c60fd6095ac2bb02c2908cf4dd7c81f11c289e4bce98f3553768f392a80ce22bf5c4f4a248c6b",
```
Comment (Thegaram) [inline core/vm/testdata/precompiles/modexp.json:17]: Maybe we can get some 32-byte test cases from Ethereum mainnet and add them here

**Review by Thegaram — APPROVED** [review]

**Review by icemelon — APPROVED** [review]

**Pull request merged**